src/util/bip341.rs

```diff
@@ -304,67 +310,50 @@ impl<'a> SigHashCache<'a> {
         Ok(TapSighashHash::from_engine(enc))
     }
 
-    /// Calculate hash for prevouts scriptpubkeys
-    pub fn hash_scriptpubkeys(&mut self, prevouts: &[TxOut]) -> sha256::Hash {
-        let hash_scriptpubkeys = &mut self.hash_scriptpubkeys;
-        *hash_scriptpubkeys.get_or_insert_with(|| {
-            let mut enc = sha256::Hash::engine();
-            for prevout in prevouts {
-                prevout.script_pubkey.consensus_encode(&mut enc).unwrap();
+    fn common_cache(&mut self) -> &CommonCache {
+        if self.common_cache.is_none() {
+            let mut enc_prevouts = sha256::Hash::engine();
+            let mut enc_sequences = sha256::Hash::engine();
+            for txin in self.tx.input.iter() {
+                txin.previous_output
+                    .consensus_encode(&mut enc_prevouts)
+                    .unwrap();
+                txin.sequence.consensus_encode(&mut enc_sequences).unwrap();
             }
-            sha256::Hash::from_engine(enc)
-        })
+            let cache = CommonCache {
+                prevouts: sha256::Hash::from_engine(enc_prevouts),
+                sequences: sha256::Hash::from_engine(enc_sequences),
+                outputs: {
+                    let mut enc = sha256::Hash::engine();
+                    for txout in self.tx.output.iter() {
+                        txout.consensus_encode(&mut enc).unwrap();
+                    }
+                    sha256::Hash::from_engine(enc)
+                },
+            };
+            self.common_cache = Some(cache);
+        }
+        self.common_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
```

Comment:
I changed it for the following reasons:

* the `segwit_cache()` method is now calling the `common_cache()` method that takes `&self`, so it's not possible anymore to borrow and mutable borrow at the same time (before was possible because we were borrowing and mutable borrowing on different struct 34fields).
* `common_cache()` and `taproot_cache()` could indeed use  `get_or_insert_with()` but it requires let bindings before the call that are done even in the cache hit phase, so I think it's a little better to avoid those (unless someone tells me they are optimized away by the compiler).
* since  it seems `segwit_cache()` cannot be implemented with `get_or_insert_with()`, I thought to be coherent with other methods

Unrelated to the `get_or_insert_with()` discussion, the method before was returning owned values, copying 32 bytes every time, while now it returns only a borrow

Reply:
I really don't believe there's performance loss. It is also possible to use `get_or_insert_with` in `segwit_cache` with some refactoring - see last commit of https://github.com/RCasatta/rust-bitcoin/pull/3